md/groovy-in-bonita.md

```diff
@@ -0,0 +1,359 @@
+
+```groovy
+def list = ['Daffy', 'Bugs', 'Elmer', 'Tweety', 'Silvester', 'Yosemite']
+assert 'Bugs' == list.find { it == 'Bugs' }
+assert ['Daffy', 'Bugs', 'Elmer'] == list.findAll { it.size() < 6 }
+assert list.any { it =~ /a/ }
+assert list.every { it.size() > 3 }
+ 
+def map = [name: 'Messages from mrhaki', url: 'http://mrhaki.blogspot.com', blog: true]
+def found = map.find { key, value -> key == 'name' }
+assert found.key == 'name' && found.value == 'Messages from mrhaki'
+found = map.find { it.value =~ /mrhaki/ }
+assert found.key == 'name' && found.value == 'Messages from mrhaki'
+assert [name: 'Messages from mrhaki', url: 'http://mrhaki.blogspot.com'] == map.findAll { key, value -> value =~ /mrhaki/ }
+assert map.any { entry -> entry.value }
+assert map.every { key, value -> key.size() >= 3 }
+```
+
+More about Closures in Groovy [here](http://docs.groovy-lang.org/docs/groovy-2.4.16/html/documentation/#_closures).
+
+### Operators
+
+* Identity operator
+
+In Groovy, using `==` to test equality is different from using the same operator in Java. In Groovy, it is calling equals. If you want to compare reference equality, you should use is like in the following example:
+
+```groovy
+def list1 = ['Groovy 1.8','Groovy 2.0','Groovy 2.3'] // Create a list of strings   
+def list2 = ['Groovy 1.8','Groovy 2.0','Groovy 2.3'] // Create another list of strings containing the same elements       
+assert list1 == list2 // using ==, we test object equality                                 
+assert !list1.is(list2) //	but using 'is', we can check that references are distinct
+```
+
+
+* Safe navigation operator `?.` to Avoid NullPointerException:
+
+Suppose we have a simple model like this:
+
+```groovy
+class Company {
+    Address address
+    String name
+}
+ 
+class Address {
+    Street street
+    String postalCode
+    String city
+}
+ 
+class Street {
+    String name
+    String number
+    String additionalInfo
+}
+```
+
+We want to display the street name, but we don't know if all object instances are available. To avoid a NullPointerException we write the following code:
+
+```groovy
+// company can be null.
+if (company != null && company.getAddress() != null && company.getAddress().getStreet() != null) {
+    println company.address.street.name
+}
+```
+
+Groovy adds the safe navigation operator to shorten all this to:
+
+```groovy
+// company can be null.
+println company?.address?.street?.name
+```
+
+* The Elvis operator `:?` to shorten ternary expression
+
+```groovy
+def sampleText
+ 
+// Normal ternary operator.
+def ternaryOutput = (sampleText != null) ? sampleText : 'Hello Groovy!'
+ 
+// The Elvis operator in action. We must read: 'If sampleText is not null assign
+// sampleText to elvisOuput, otherwise assign 'Viva Las Vegas!' to elvisOutput.
+def elvisOutput = sampleText ?: 'Viva Las Vegas!'
+```
+
+### Groovy truth
+
+Groovy decides whether an expression is true or false by applying the rules given below.  
```

Comment:
the following rules doesn't apply when using the null safe operator or the elvis operator, those operators only check if the reference is null or not.

Reply:
I don't understand what you mean ?

Reply:
``` groovy
def emptyString = ""
println emptyString?.toLong()
// NumberFormatException -> the check on the String doesn't return false, toLong() is performed.
```

Reply:
It is a border use case.